swabbie-aws/src/main/kotlin/com/netflix/spinnaker/swabbie/aws/AWS.kt

```diff
@@ -40,3 +88,398 @@ interface AWS {
   fun getServerGroups(params: Parameters): List<AmazonAutoScalingGroup>
   fun getServerGroup(params: Parameters): AmazonAutoScalingGroup?
 }
+
+/**
+ *  An AWS api proxy
+ */
+class Proxy<T : AWS>(
+  val api: T
+)
+
+/**
+ * AWS api impl
+ * @property proxy calls will be forwarded to supplied proxy
+ */
+class Vanilla(
+  private val sts: AWSSecurityTokenService,
+  private val proxy: Optional<Proxy<AWS>>,
+  private val objectMapper: ObjectMapper,
+  accountProvider: AccountProvider
+) : AWS {
+  private val log: Logger = LoggerFactory.getLogger(javaClass)
+
+  init {
+    if (proxy.isPresent) {
+      log.info("Using aws proxy {}", proxy.javaClass.name)
+    }
+  }
+
+  override fun getInstances(params: Parameters): List<AmazonInstance> {
+    if (proxy.isPresent) {
```

Comment:
I don't love that every function here has this `if (proxy.isPresent) {` bit which does the same thing. Since it looks like if there _is_ a proxy it is applicable for every call, can we make this cleaner? We can discuss offline on ways to make this nicer.